### PR TITLE
Multilevel ibfe 5

### PR DESCRIPTION
**Pull request by drwells**
This is enough to get velocity interpolation working correctly - we get about the same (they are a factor of 2 better) errors as we do when the finest grid level is one level lower and all the elements are on that level.

`FEDataManager::collectActivePatchElements()` is getting a bit messy but I think it would be best for now to leave refactoring that for later - this is already pretty complicated.

### IBAMR Pull Request Checklist
- [x] Does the test suite pass?
- [x] Was clang-format run?
- [x] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [x] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [x] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [x] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [x] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

### CODE REVIEW

File: ibtk/include/ibtk/FEDataManager.h [hunk 1]
```diff
@@ -294,9 +294,25 @@ class SubdomainToLevelTranslation
         return it->second;
     }
 
+    /*!
+     * Return whether or not there are any elements on a given level.
+     */
+    bool levelHasElements(const int level_number) const
```
Comment (boyceg) [inline ibtk/include/ibtk/FEDataManager.h:300]: 👍

File: ibtk/src/lagrangian/FEDataManager.cpp [hunk 2]
```diff
@@ -2543,7 +2531,7 @@ FEDataManager::applyGradientDetector(const Pointer<BasePatchHierarchy<NDIM> > hi
         // level.
         std::vector<std::vector<Elem*> > active_level_elem_map;
         const IntVector<NDIM> ghost_width = 1;
-        collectActivePatchElements(active_level_elem_map, level_number);
+        collectActivePatchElements(active_level_elem_map, level_number, level_number + 1, d_max_level_number);
```
Comment (drwells) [inline ibtk/src/lagrangian/FEDataManager.cpp:2534]: This more precise API call replaces `if (getPatchLevel(elem) <= level_number) continue;`

File: ibtk/src/lagrangian/FEDataManager.cpp [hunk 3]
```diff
@@ -564,21 +564,9 @@ FEDataManager::reinitElementMappings()
...
-        d_active_patch_elem_map[ln].resize(level->getProcessorMapping().getNumberOfLocalIndices());
-        d_active_patch_node_map[ln].resize(level->getProcessorMapping().getNumberOfLocalIndices());
-    }
-
-    // TODO: collectActivePatchElements still assumes every element is on the
-    // same patch level.
-    collectActivePatchElements(d_active_patch_elem_map.back(), d_max_level_number);
     for (int ln = 0; ln <= d_max_level_number; ++ln)
     {
+        collectActivePatchElements(d_active_patch_elem_map[ln], ln, ln, ln);
```
Comment (drwells) [inline ibtk/src/lagrangian/FEDataManager.cpp:569]: This is the main change in the patch - we now actually assign elements to their requested levels.

File: ibtk/include/ibtk/FEDataManager.h (hunk 1 above)
Comment (drwells) [reply ibtk/include/ibtk/FEDataManager.h:300]: Unless there is a bug in the implementation the check that uses this should prevent us from calculating the global set of bounding boxes any more times than we currently do on `master`.

**Comment by drwells** [comment]
@boyceg are you happy with this? If so lets merge so we can move on to spreading :)

**Comment by boyceg** [comment]
I got sidetracked. I will check it out tonight.

File: ibtk/include/ibtk/FEDataManager.h [hunk 4]
```diff
@@ -1129,10 +1145,29 @@ class FEDataManager : public SAMRAI::tbox::Serializable
      * are allowed to move no more than one cell width between regridding
      * operations).
      *
+     * The parameters refer to the levels of different objects:
+     * <ol>
+     *   <li>@p level_number - the level number in the patch hierarchy on which
+     *     we are identifying intersections.</li>
+     *   <li>@p coarsest_elem_ln - The minimum level number of elements we should
```
Comment (boyceg) [inline ibtk/include/ibtk/FEDataManager.h:1152]: For input parameters, I wonder if it is better to be more verbose, e.g.:
```suggestion
     *   <li>@p coarsest_elem_level_num - The minimum level number of elements we should
```

File: ibtk/include/ibtk/FEDataManager.h [hunk 5]
```diff
@@ -1129,10 +1145,29 @@ class FEDataManager : public SAMRAI::tbox::Serializable
...
+     *     consider (see the main documentation of this class for an explanation
+     *     on how elements are assigned to particular levels)</li>
+     *   <li>@p finest_elem_ln - The maximum level number of elements we should
```
Comment (boyceg) [inline ibtk/include/ibtk/FEDataManager.h:1155]: Similarly:
```suggestion
     *   <li>@p finest_elem_level_num - The maximum level number of elements we should
```

File: ibtk/include/ibtk/FEDataManager.h [hunk 6]
```diff
@@ -1129,10 +1145,29 @@ class FEDataManager : public SAMRAI::tbox::Serializable
...
+     * elements on levels higher than the present level) and to do IB
+     * calculations (where all three numbers will be the same).
+     *
      * In this method, the determination as to whether an element is local or
      * not is based on the position of the bounding box of the element.
      */
-    void collectActivePatchElements(std::vector<std::vector<libMesh::Elem*> >& active_patch_elems, int level_number);
+    void collectActivePatchElements(std::vector<std::vector<libMesh::Elem*> >& active_patch_elems,
+                                    int level_number,
+                                    int coarsest_elem_ln,
```
Comment (boyceg) [inline ibtk/include/ibtk/FEDataManager.h:1169]:
```suggestion
                                    int coarsest_elem_level_num,
```

File: ibtk/include/ibtk/FEDataManager.h [hunk 7]
```diff
@@ -1129,10 +1145,29 @@ class FEDataManager : public SAMRAI::tbox::Serializable
...
+                                    int finest_elem_ln);
```
Comment (boyceg) [inline ibtk/include/ibtk/FEDataManager.h:1170]:
```suggestion
                                    int finest_elem_level_num);
```
...although I guess this is in the private interface, so maybe it does not matter.

**Review by boyceg — APPROVED** [review]
Looks good to me.

**Comment by drwells** [comment]
I went with `_ln` since that is what we use in `updateQuadPointCountData` and a couple of other functions - should we use `level_num` everywhere in the future?

**Comment by drwells** [comment]
Since this is approved I will merge to keep the ball rolling and open an issue to track style changes.

**Comment by boyceg** [comment]
> I went with `_ln` since that is what we use in `updateQuadPointCountData` and a couple of other functions - should we use `level_num` everywhere in the future?

I think `_ln` is fine. It may make sense to standardize on a suffix at some point.

**Comment by boyceg** [comment]
I can't wait for _Multilevel ibfe 6_. 😀

**Comment by drwells** [comment]
I didn't have to fix any bugs in `interpWeighted` for this to work so I am cautiously hopeful that multilevel spreading won't be too bad.

**Pull request merged**